File: main/src/controllers/seller.controller.ts
import { Request, Response, NextFunction } from "express";
import { SellerService } from "../services/seller.service.js";

export class SellerController {
  // 1. Dashboard
  static getDashboard(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const stats = SellerService.getDashboardStats(sellerId);
      res.json({ success: true, data: stats });
    } catch (err) {
      next(err);
    }
  }

  // 2. Shop Profile
  static getShop(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const shop = SellerService.getShopProfile(sellerId);
      res.json({ success: true, data: shop });
    } catch (err) {
      next(err);
    }
  }

  static updateShop(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const updated = SellerService.updateShopProfile(sellerId, req.body);
      res.json({ success: true, message: "Cập nhật hồ sơ cửa hàng thành công!", data: updated });
    } catch (err) {
      next(err);
    }
  }

  // 3. Products
  static getProducts(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const filters = {
        search: req.query.search as string,
        categoryId: req.query.categoryId ? Number(req.query.categoryId) : undefined,
        status: req.query.status as string
      };
      const products = SellerService.getProducts(sellerId, filters);
      res.json({ success: true, count: products.length, data: products });
    } catch (err) {
      next(err);
    }
  }

  static getProductById(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const productId = Number(req.params.id);
      const product = SellerService.getProductById(sellerId, productId);
      if (!product) {
        res.status(404).json({ success: false, error: "Không tìm thấy sản phẩm." });
        return;
      }
      res.json({ success: true, data: product });
    } catch (err) {
      next(err);
    }
  }

  static createProduct(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const created = SellerService.createProduct(sellerId, req.body);
      res.status(201).json({ success: true, message: "Thêm sản phẩm mới thành công!", data: created });
    } catch (err) {
      next(err);
    }
  }

  static updateProduct(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const productId = Number(req.params.id);
      const updated = SellerService.updateProduct(sellerId, productId, req.body);
      res.json({ success: true, message: "Cập nhật sản phẩm thành công!", data: updated });
    } catch (err) {
      next(err);
    }
  }

  static deleteProduct(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const productId = Number(req.params.id);
      SellerService.deleteProduct(sellerId, productId);
      res.json({ success: true, message: "Đã lưu trữ sản phẩm thành công." });
    } catch (err) {
      next(err);
    }
  }

  // 4. Inventory
  static getInventory(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const inventory = SellerService.getInventory(sellerId);
      res.json({ success: true, data: inventory });
    } catch (err) {
      next(err);
    }
  }

  static updateStock(req: Request, res: Response, next: NextFunction) {
    try {
      const productId = Number(req.params.id);
      const { quantity, lowStockThreshold } = req.body;
      SellerService.updateStock(productId, Number(quantity), lowStockThreshold !== undefined ? Number(lowStockThreshold) : 5);
      res.json({ success: true, message: "Cập nhật số lượng tồn kho thành công!" });
    } catch (err) {
      next(err);
    }
  }

  // 5. Orders
  static getOrders(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const status = req.query.status as string;
      const orders = SellerService.getOrders(sellerId, status);
      res.json({ success: true, count: orders.length, data: orders });
    } catch (err) {
      next(err);
    }
  }

  static getOrderDetail(req: Request, res: Response, next: NextFunction) {
    try {
      const orderId = Number(req.params.id);
      const order = SellerService.getOrderDetail(orderId);
      if (!order) {
        res.status(404).json({ success: false, error: "Không tìm thấy đơn hàng." });
        return;
      }
      res.json({ success: true, data: order });
    } catch (err) {
      next(err);
    }
  }

  static updateOrderStatus(req: Request, res: Response, next: NextFunction) {
    try {
      const orderId = Number(req.params.id);
      const { status } = req.body;
      SellerService.updateOrderStatus(orderId, status);
      res.json({ success: true, message: `Đã cập nhật trạng thái đơn hàng sang ${status}` });
    } catch (err) {
      next(err);
    }
  }

  // 6. Shipping Channels
  static getShipping(req: Request, res: Response, next: NextFunction) {
    try {
      const channels = SellerService.getShippingChannels();
      res.json({ success: true, data: channels });
    } catch (err) {
      next(err);
    }
  }

  static toggleShipping(req: Request, res: Response, next: NextFunction) {
    try {
      const id = Number(req.params.id);
      const { isEnabled } = req.body;
      SellerService.toggleShippingChannel(id, Boolean(isEnabled));
      res.json({ success: true, message: "Cập nhật kênh vận chuyển thành công!" });
    } catch (err) {
      next(err);
    }
  }

  // 7. Promotions
  static getPromotions(req: Request, res: Response, next: NextFunction) {
    try {
      const shopId = 1;
      const promotions = SellerService.getPromotions(shopId);
      res.json({ success: true, data: promotions });
    } catch (err) {
      next(err);
    }
  }

  static createPromotion(req: Request, res: Response, next: NextFunction) {
    try {
      const shopId = 1;
      const promo = SellerService.createPromotion(shopId, req.body);
      res.status(201).json({ success: true, message: "Tạo mã khuyến mãi thành công!", data: promo });
    } catch (err) {
      next(err);
    }
  }

  static togglePromotion(req: Request, res: Response, next: NextFunction) {
    try {
      const id = Number(req.params.id);
      const { isActive } = req.body;
      SellerService.togglePromotion(id, Boolean(isActive));
      res.json({ success: true, message: "Cập nhật trạng thái voucher thành công!" });
    } catch (err) {
      next(err);
    }
  }

  // 8. Finance / Wallet
  static getWallet(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const wallet = SellerService.getWallet(sellerId);
      res.json({ success: true, data: wallet });
    } catch (err) {
      next(err);
    }
  }

  static requestWithdraw(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const { amount, note } = req.body;
      const wallet = SellerService.requestWithdraw(sellerId, Number(amount), note);
      res.json({ success: true, message: "Yêu cầu rút tiền thành công!", data: wallet });
    } catch (err) {
      next(err);
    }
  }

  // 9. Analytics
  static getAnalytics(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const analytics = SellerService.getAnalytics(sellerId);
      res.json({ success: true, data: analytics });
    } catch (err) {
      next(err);
    }
  }

  // 10. Reviews
  static getReviews(req: Request, res: Response, next: NextFunction) {
    try {
      const sellerId = (req as any).user?.id || 1;
      const reviews = SellerService.getReviews(sellerId);
      res.json({ success: true, data: reviews });
    } catch (err) {
      next(err);
    }
  }

  static replyReview(req: Request, res: Response, next: NextFunction) {
    try {
      const reviewId = Number(req.params.id);
      const { reply } = req.body;
      SellerService.replyReview(reviewId, reply);
      res.json({ success: true, message: "Đã gửi phản hồi đánh giá thành công!" });
    } catch (err) {
      next(err);
    }
  }

  // 11. AI Description
  static generateAIDescription(req: Request, res: Response, next: NextFunction) {
    try {
      const aiResult = SellerService.generateAIDescription(req.body);
      res.json({ success: true, data: aiResult });
    } catch (err) {
      next(err);
    }
  }
}

File: main/src/DM_FashionMall.API/Controllers/SellerController.cs
using System;
using System.Security.Claims;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using DM_FashionMall.API.DTOs;
using DM_FashionMall.API.Services;

namespace DM_FashionMall.API.Controllers
{
    [ApiController]
    [Route("api/seller")]
    public class SellerController : ControllerBase
    {
        private readonly ISellerService _sellerService;

        public SellerController(ISellerService sellerService)
        {
            _sellerService = sellerService;
        }

        private int GetCurrentShopId()
        {
            if (Request.Headers.TryGetValue("X-Shop-Id", out var headerVal) && int.TryParse(headerVal, out int hId))
            {
                return hId;
            }

            var shopClaim = User.FindFirst("ShopId")?.Value;
            if (int.TryParse(shopClaim, out int sId) && sId > 0)
            {
                return sId;
            }

            return 1; // Default sample shop ID
        }

        private int GetCurrentUserId()
        {
            var userClaim = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
            if (int.TryParse(userClaim, out int uId) && uId > 0)
            {
                return uId;
            }

            return 1; // Default sample user ID
        }

        // 2. Dashboard
        [HttpGet("dashboard/metrics")]
        public async Task<IActionResult> GetDashboardMetrics()
        {
            var metrics = await _sellerService.GetDashboardMetricsAsync(GetCurrentShopId());
            return Ok(metrics);
        }

        [HttpGet("dashboard/chart")]
        public async Task<IActionResult> GetSalesChart([FromQuery] int days = 7)
        {
            var data = await _sellerService.GetSalesChartDataAsync(GetCurrentShopId(), days);
            return Ok(data);
        }

        // 3. Shop Profile
        [HttpGet("profile")]
        public async Task<IActionResult> GetProfile()
        {
            var profile = await _sellerService.GetShopProfileAsync(GetCurrentShopId());
            if (profile == null) return NotFound(new { message = "Không tìm thấy hồ sơ shop" });
            return Ok(profile);
        }

        [HttpPut("profile")]
        public async Task<IActionResult> UpdateProfile([FromBody] UpdateShopProfileRequest request)
        {
            var success = await _sellerService.UpdateShopProfileAsync(GetCurrentShopId(), request);
            if (!success) return BadRequest(new { message = "Cập nhật hồ sơ shop thất bại" });
            return Ok(new { message = "Cập nhật hồ sơ shop thành công" });
        }

        // 4, 5, 6. Products
        [HttpGet("products")]
        public async Task<IActionResult> GetProducts([FromQuery] string? search, [FromQuery] int? categoryId, [FromQuery] string? status)
        {
            var list = await _sellerService.GetProductsAsync(GetCurrentShopId(), search, categoryId, status);
            return Ok(list);
        }

        [HttpGet("products/{id}")]
        public async Task<IActionResult> GetProduct(int id)
        {
            var prod = await _sellerService.GetProductDetailAsync(GetCurrentShopId(), id);
            if (prod == null) return NotFound(new { message = "Không tìm thấy sản phẩm" });
            return Ok(prod);
        }

        [HttpPost("products")]
        public async Task<IActionResult> CreateProduct([FromBody] SaveProductRequest request)
        {
            var created = await _sellerService.CreateProductAsync(GetCurrentUserId(), GetCurrentShopId(), request);
            return CreatedAtAction(nameof(GetProduct), new { id = created.Id }, created);
        }

        [HttpPut("products/{id}")]
        public async Task<IActionResult> UpdateProduct(int id, [FromBody] SaveProductRequest request)
        {
            var success = await _sellerService.UpdateProductAsync(GetCurrentShopId(), id, request);
            if (!success) return NotFound(new { message = "Không tìm thấy sản phẩm để cập nhật" });
            return Ok(new { message = "Cập nhật sản phẩm thành công" });
        }

        [HttpDelete("products/{id}")]
        public async Task<IActionResult> DeleteProduct(int id)
        {
            var success = await _sellerService.DeleteProductAsync(GetCurrentShopId(), id);
            if (!success) return NotFound(new { message = "Không tìm thấy sản phẩm" });
            return Ok(new { message = "Xóa sản phẩm thành công" });
        }

        [HttpPatch("products/{id}/status")]
        public async Task<IActionResult> ToggleProductStatus(int id)
        {
            var success = await _sellerService.ToggleProductStatusAsync(GetCurrentShopId(), id);
            if (!success) return NotFound(new { message = "Không tìm thấy sản phẩm" });
            return Ok(new { message = "Chuyển trạng thái sản phẩm thành công" });
        }
        // 7. Inventory
        [HttpGet("inventory")]
        public async Task<IActionResult> GetInventory([FromQuery] bool lowStockOnly = false)
        {
            var list = await _sellerService.GetInventoryAsync(GetCurrentShopId(), lowStockOnly);
            return Ok(list);
        }

        [HttpPatch("inventory/quick-update")]
        public async Task<IActionResult> QuickUpdateStock([FromBody] QuickUpdateStockRequest request)
        {
            var success = await _sellerService.QuickUpdateStockAsync(GetCurrentShopId(), request);
            if (!success) return NotFound(new { message = "Không tìm thấy sản phẩm/biến thể để cập nhật kho" });
            return Ok(new { message = "Cập nhật tồn kho thành công" });
        }

        // 8, 9. Orders
        [HttpGet("orders")]
        public async Task<IActionResult> GetOrders([FromQuery] string? status = null)
        {
            var list = await _sellerService.GetOrdersAsync(GetCurrentShopId(), status);
            return Ok(list);
        }

        [HttpGet("orders/{id}")]
        public async Task<IActionResult> GetOrderDetail(int id)
        {
            var order = await _sellerService.GetOrderDetailAsync(GetCurrentShopId(), id);
            if (order == null) return NotFound(new { message = "Không tìm thấy đơn hàng" });
            return Ok(order);
        }

        [HttpPut("orders/{id}/status")]
        public async Task<IActionResult> UpdateOrderStatus(int id, [FromBody] UpdateOrderStatusRequest request)
        {
            var success = await _sellerService.UpdateOrderStatusAsync(GetCurrentShopId(), id, request);
            if (!success) return NotFound(new { message = "Không tìm thấy đơn hàng để cập nhật trạng thái" });
            return Ok(new { message = $"Đã cập nhật trạng thái đơn hàng sang {request.Status}" });
        }

        // 10. Shipping
        [HttpGet("shipping/channels")]
        public async Task<IActionResult> GetShippingChannels()
        {
            var channels = await _sellerService.GetShippingChannelsAsync();
            return Ok(channels);
        }

        [HttpPost("shipping/channels/{id}/toggle")]
        public async Task<IActionResult> ToggleShippingChannel(int id)
        {
            var success = await _sellerService.ToggleShippingChannelAsync(id);
            if (!success) return NotFound(new { message = "Không tìm thấy kênh vận chuyển" });
            return Ok(new { message = "Đã cập nhật trạng thái kênh vận chuyển" });
        }

        // 11. Promotions
        [HttpGet("promotions")]
        public async Task<IActionResult> GetPromotions()
        {
            var list = await _sellerService.GetPromotionsAsync(GetCurrentShopId());
            return Ok(list);
        }

        [HttpPost("promotions")]
        public async Task<IActionResult> CreatePromotion([FromBody] CreatePromotionRequest request)
        {
            var promo = await _sellerService.CreatePromotionAsync(GetCurrentShopId(), request);
            return Ok(promo);
        }

        [HttpDelete("promotions/{id}")]
        public async Task<IActionResult> DeletePromotion(int id)
        {
            var success = await _sellerService.DeletePromotionAsync(GetCurrentShopId(), id);
            if (!success) return NotFound(new { message = "Không tìm thấy mã khuyến mãi" });
            return Ok(new { message = "Xóa mã khuyến mãi thành công" });
        }

        // 12. Finance
        [HttpGet("finance/wallet")]
        public async Task<IActionResult> GetWallet()
        {
            var wallet = await _sellerService.GetWalletAsync(GetCurrentUserId());
            if (wallet == null) return NotFound(new { message = "Không tìm thấy ví người bán" });
            return Ok(wallet);
        }

        [HttpPost("finance/withdraw")]
        public async Task<IActionResult> Withdraw([FromBody] WithdrawRequest request)
        {
            var success = await _sellerService.RequestWithdrawAsync(GetCurrentUserId(), request);
            if (!success) return BadRequest(new { message = "Số dư không đủ hoặc thông tin ngân hàng không hợp lệ" });
            return Ok(new { message = "Lệnh rút tiền đã được thực hiện thành công" });
        }

        // 13. Analytics
        [HttpGet("analytics/overview")]
        public async Task<IActionResult> GetAnalyticsOverview()
        {
            var data = await _sellerService.GetAnalyticsOverviewAsync(GetCurrentShopId());
            return Ok(data);
        }

        // 14. Reviews
        [HttpGet("reviews")]
        public async Task<IActionResult> GetReviews([FromQuery] int? rating = null)
        {
            var list = await _sellerService.GetReviewsAsync(GetCurrentShopId(), rating);
            return Ok(list);
        }

        [HttpPost("reviews/{id}/reply")]
        public async Task<IActionResult> ReplyReview(int id, [FromBody] ReplyReviewRequest request)
        {
            var success = await _sellerService.ReplyReviewAsync(GetCurrentShopId(), id, request);
            if (!success) return NotFound(new { message = "Không tìm thấy đánh giá" });
            return Ok(new { message = "Đã gửi phản hồi đánh giá thành công" });
        }

        // 15. Settings
        [HttpGet("settings")]
        public async Task<IActionResult> GetSettings()
        {
            var settings = await _sellerService.GetSettingsAsync(GetCurrentShopId());
            return Ok(settings);
        }

        [HttpPut("settings")]
        public async Task<IActionResult> UpdateSettings([FromBody] UpdateSettingsRequest request)
        {
            var success = await _sellerService.UpdateSettingsAsync(GetCurrentShopId(), request);
            if (!success) return BadRequest(new { message = "Cập nhật cài đặt thất bại" });
            return Ok(new { message = "Cập nhật cài đặt thành công" });
        }

        [HttpPost("settings/change-password")]
        public async Task<IActionResult> ChangePassword([FromBody] ChangePasswordRequest request)
        {
            var success = await _sellerService.ChangePasswordAsync(GetCurrentUserId(), request);
            if (!success) return BadRequest(new { message = "Mật khẩu hiện tại không chính xác" });
            return Ok(new { message = "Đổi mật khẩu thành công" });
        }

        // AI Copywriter
        [HttpPost("ai/generate-description")]
        public async Task<IActionResult> GenerateDescription([FromBody] AiGenerateDescriptionRequest request)
        {
            var result = await _sellerService.GenerateProductDescriptionAsync(request);
            return Ok(result);
        }

        // Helpers
        [HttpGet("categories")]
        public async Task<IActionResult> GetCategories()
        {
            var cats = await _sellerService.GetCategoriesAsync();
            return Ok(cats);
        }

        [HttpGet("brands")]
        public async Task<IActionResult> GetBrands()
        {
            var brands = await _sellerService.GetBrandsAsync();
            return Ok(brands);
        }
    }
}
